launch_frontend/launch_frontend/parser.py

```diff
@@ -0,0 +1,113 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for Parser class and parsing methods."""
+
+import io
+from typing import Text
+from typing import Union
+
+import launch
+from launch.utilities import is_a
+
+from pkg_resources import iter_entry_points
+
+from .entity import Entity
+from .expose import action_parse_methods
+from .parse_substitution import default_parse_substitution
+
+interpolation_fuctions = {
+    entry_point.name: entry_point.load()
+    for entry_point in iter_entry_points('launch_frontend.interpolate_substitution')
+}
+
+extensions_loaded = False
+
+
+class Parser:
+    """
+    Abstract class for parsing actions, substitutions and descriptions.
+
+    Implementations of the parser class, should override the load method.
+    They could also override the parse_substitution method, or not.
+    load_parser_extensions, parse_action and parse_description are not suposed to be overrided.
+    """
+
+    extensions_loaded = False
+    frontend_parsers = None
+
+    @classmethod
+    def load_parser_extensions(cls):
+        """Load parser extension, in order to get all the exposed substitutions and actions."""
+        if cls.extensions_loaded is False:
+            for entry_point in iter_entry_points('launch_frontend.parser_extension'):
+                entry_point.load()
+            cls.extensions_loaded = True
+
+    @classmethod
+    def load_parser_implementations(cls):
+        """Load all the available frontend entities."""
+        if cls.frontend_parsers is None:
+            cls.frontend_parsers = {
+                entry_point.name: entry_point.load()
+                for entry_point in iter_entry_points('launch_frontend.parser')
+            }
+
+    @classmethod
+    def parse_action(cls, entity: Entity) -> launch.Action:
+        """Parse an action, using its registered parsing method."""
+        cls.load_parser_extensions()
+        if entity.type_name not in action_parse_methods:
+            raise RuntimeError('Unrecognized entity of the type: {}'.format(entity.type_name))
+        action, kwargs = action_parse_methods[entity.type_name](entity, cls)
+        return action(**kwargs)
+
+    @classmethod
+    def parse_substitution(cls, value: Text) -> launch.SomeSubstitutionsType:
+        """Parse a substitution."""
+        return default_parse_substitution(value)
+
+    @classmethod
+    def parse_description(cls, entity: Entity) -> launch.LaunchDescription:
```

Comment:
`parse_description`, `parse_action` should. `cls` is finally passed to the action parsing methods as `parser` argument. The parsing methods could call `parse.substitution`, which can be overridden by the derived class.

Reply:
Right, my point being that these should be instance methods.

Reply:
The `Parse` instances don't have a state now, but it could have it in the future, so I think it is a good idea. I will change `parse_action`, `parse_substitution`, `parse_description` to instance methods.
`load`, `load_parser_extensions` and `load_parser_implementations` will be continue being class methods.
I will stop returning an object of `Parse` type in the `load` method (ie: a subclass of `Parse`) , and start returning an instance of it.
Note: I will also rename `load_parser_extensions` as `load_launch_extensions`.